modules/internal/ChapelBase.chpl

```diff
@@ -2330,6 +2330,32 @@ module ChapelBase {
   }
   var chpl_moduleDeinitFuns = nil: unmanaged chpl_ModuleDeinit?;
 
+  // Supports type field accessors on nilable classes - on an instance...
+  inline proc chpl_checkLegalTypeFieldAccessor(thisArg, type fieldType,
```

Comment:
Why have an error here in module code? Wouldn't it be simpler / less confusing to just put the error message in the compiler, instead of adding a call to this?

Reply:
Sure, it is possible.

In the past we have offloaded pieces of work from the compiler to the modules. A big motivation for that was that Chapel code is more concise and more readable than building the corresponding AST from within the compiler.

I am not concerned about the added confusion because (a) I think it is very minor and (b) the developer would need to deal with it only in some corner cases.

I don't even know if compiler-generated AST can produce a clearer error message than calling out into Chapel code.

Recall that this happens before resolution, so we do not necessarily know (a) the type of the field and (b) the flavor of the receiver ie its nilability and memory management strategy. So I do not see how the compiler can specialize the AST any better than what we do in the module code.

What do you think?

Reply:
Thanks for reminding me it occurs before resolution, that makes me more inclined to keep the current strategy.